cda-core/src/diag_kernel/diagservices.rs

```diff
@@ -372,3 +373,131 @@ impl DiagServiceResponseStruct {
         }
     }
 }
+
+/// Extracts NRC information from raw UDS negative response data.
+/// UDS Negative Response format: [0x7F, SID, NRC]
+/// This function provides a default handler for standardized NRCs when no
+/// database-specific mapping is available.
+fn extract_nrc_from_raw_data(data: &[u8]) -> Result<MappedNRC, DiagServiceError> {
+    // Minimum length for a negative response is 3 bytes: 0x7F + SID + NRC
+    if data.len() < 3 {
+        return Err(DiagServiceError::UnexpectedResponse(Some(
+            "Negative response data too short".to_owned(),
+        )));
+    }
+
+    // Verify this is a negative response (first byte should be 0x7F)
+    if data.first() != Some(&0x7F) {
+        return Err(DiagServiceError::UnexpectedResponse(Some(
+            "Expected negative response (0x7F) but got different SID".to_owned(),
+        )));
+    }
+
+    let sid = data.get(1).copied();
+    let nrc = data.get(2).copied();
+
+    match nrc {
+        Some(nrc_code) => {
+            let description = iso_14229_nrc::get_nrc_code(nrc_code).to_owned();
+            Ok(MappedNRC {
+                code: Some(nrc_code),
+                description: Some(description),
+                sid,
+            })
+        }
+        None => Err(DiagServiceError::UnexpectedResponse(Some(
+            "Missing NRC code in negative response".to_owned(),
+        ))),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_extract_nrc_from_raw_data_valid() {
+        // Test with a valid negative response: 0x7F (neg response), 0x22 (SID), 0x33 (NRC: Security Access Denied)
+        let data = vec![0x7F, 0x22, 0x33];
+        let result = extract_nrc_from_raw_data(&data).unwrap();
+
+        assert_eq!(result.code, Some(0x33));
+        assert_eq!(result.sid, Some(0x22));
+        assert_eq!(result.description, Some("Security Access Denied".to_owned()));
+    }
+
+    #[test]
+    fn test_extract_nrc_from_raw_data_general_reject() {
+        // Test with NRC 0x10 (General Reject)
+        let data = vec![0x7F, 0x19, 0x10];
+        let result = extract_nrc_from_raw_data(&data).unwrap();
+
+        assert_eq!(result.code, Some(0x10));
+        assert_eq!(result.sid, Some(0x19));
+        assert_eq!(result.description, Some("General Reject".to_owned()));
+    }
+
+    #[test]
+    fn test_extract_nrc_from_raw_data_response_pending() {
```

Comment:
suggestion: The first three tests added here, do test pretty much the same thing with different nrcs. I would prefer only keeping one of them, as the other tests do not add any real value imho.